src/backend/distributed/transaction/transaction_management.c

```diff
@@ -154,6 +164,17 @@ CoordinatedTransactionCallback(XactEvent event, void *arg)
 	{
 		case XACT_EVENT_COMMIT:
 		{
+			/*
+			 * ERRORs thrown during XACT_EVENT_COMMIT will cause postgres to abort, at
+			 * this point enough work has been done that it's not possible to rollback.
+			 *
+			 * One possible source of errors is memory allocation failures. To minimize
+			 * the chance of those happening we've pre-allocated some memory in the
+			 * CommitContext, it has 8kb of memory that we're allowed to use.
+			 */
+			MemoryContext previousContext = CurrentMemoryContext;
+			MemoryContextSwitchTo(CommitContext);
```

Comment:
should we also be doing this in the abort handler then?

Reply:
No, because postgres [already sets aside memory](https://github.com/postgres/postgres/blob/master/src/backend/access/transam/xact.c#L990) for abort handling [and switches to it](https://github.com/postgres/postgres/blob/master/src/backend/access/transam/xact.c#L2454) when processing an abort.

Reply:
Ok, but why _just_ do this in `EVENT_COMMIT` and not also the other blocks here?

FWIW I think this fixes the problem from my testing. I can't give you repro instructions easily though (it'll be easier after Monday).

Reply:
(I guess you're going to say that in all other cases it's OK to roll back, but maybe add that detail in the comment for future readers)